byway/src/main/java/com/google/sps/data/UserInfo.java

```diff
@@ -94,4 +102,38 @@ public static UserInfo fromEntity(Entity userInfoEntity) {
             "User entity does not contain trip Ids");
     return new UserInfo(email, userId, tripIds);
   }
+
+  public Entity toEntity() {
+    Entity userEntity = new Entity(this.getKey());
+    userEntity.setProperty("email", this.email);
+    userEntity.setProperty("tripIds", this.tripIds);
+    return userEntity;
+  }
+
+  /**
+   * Finds a User or creates User if User doesnt exist
+   *
+   * @param userService UserServiceFactory.getUserService()
+   * @param datastore DatastoreServiceFactory.getDatastoreService()
+   */
+  public static UserInfo findOrCreateUser(UserService userService, DatastoreService datastore) {
+    User user = userService.getCurrentUser();
+    // Create a key based on the user ID
+    if (user == null) {
+      return null;
+    }
+    Key userKey = KeyFactory.createKey(UserInfo.DATASTORE_ENTITY_KIND, user.getUserId());
+    UserInfo userInfo;
+    Entity userEntity;
+    try {
+      // try to retrieve the entity with the key
+      userEntity = datastore.get(userKey);
+    } catch (EntityNotFoundException exception) {
+      // If the user doesn't exist yet or is new, create a new user
+      UserInfo newUser = new UserInfo(user.getEmail(), user.getUserId(), new ArrayList<String>());
+      userEntity = newUser.toEntity();
+      datastore.put(userEntity);
```

Comment:
Note: it's clearer if you just in-line the `toEntity` call in `datastore.put`, and `fromEntity` around `datastore.get`. The code becomes shorter, and it's easier to read because you don't have to follow what's being put in the two variables declared before the `try` block, plus then you're not constructing `UserInfo` objects twice.

I would structure it something like:

```java
    Key userKey = KeyFactory.createKey(UserInfo.DATASTORE_ENTITY_KIND, user.getUserId());
    try {
      // try to retrieve the entity with the key
      return fromEntity(datastore.get(userKey));
    } catch (EntityNotFoundException exception) {
      // If the user doesn't exist yet or is new, create a new user
      UserInfo newUser = new UserInfo(user.getEmail(), user.getUserId(), new ArrayList<String>());
      datastore.put(newUser.toEntity());
      return newUser;
    }
```

I know I said to keep try-blocks as narrow as possible, but: you know that `fromEntity` does not throw `EntityNotFound` exception, and it's OK to be a bit more lax on rules like that when you are working with functions defined in the same class. A lot of the 'rules' we use are more like 'guidelines', and are intended to prevent situations where future code changes lead to unexpected behavior. Part of that is keeping in mind the distance between parts of the code that are interacting: if you are making assumptions about the behavior of code in distant parts of the codebase, it's very easy to assume something about the code there which will be changed in the future, whereas if the code you are making assumptions about is in the same class, it's often safe to assume/require that it plays nice with what you're doing. So, basically, it's fine to assume that `fromEntity` won't throw `EntityNotFound` exception (and therefore just in-line the call and return).

Reply:
fixed it